### PR TITLE
HBASE-28769 Create PoC for RangeStoreFileReader to support multi regi…

### PR DESCRIPTION
**Pull request by chrajeshbabu**
…on splitting

### CODE REVIEW

**Comment by chrajeshbabu** [comment]
Created initial range store file reader. Just raised PR for quick glance. Will add test cases and update the PR.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 33s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  |  Patch does not have any anti-patterns.  |
|||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m 14s |  |  master passed  |
| +1 :green_heart: |  compile  |   3m  0s |  |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 36s |  |  master passed  |
| +1 :green_heart: |  spotbugs  |   1m 35s |  |  master passed  |
| +1 :green_heart: |  spotless  |   0m 45s |  |  branch has no errors when running spotless:check.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m  0s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   3m  0s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   3m  0s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   0m 36s | [/results-checkstyle-hbase-server.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6141/1/artifact/yetus-general-check/output/results-checkstyle-hbase-server.txt) |  hbase-server: The patch generated 8 new + 0 unchanged - 0 fixed = 8 total (was 0)  |
| +1 :green_heart: |  spotbugs  |   1m 40s |  |  the patch passed  |
| +1 :green_heart: |  hadoopcheck  |  11m 34s |  |  Patch does not cause any errors with Hadoop 3.3.6 3.4.0.  |
| -1 :x: |  spotless  |   0m 37s |  |  patch has 63 errors when running spotless:check, run spotless:apply to fix.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 11s |  |  The patch does not generate ASF License warnings.  |
|  |   |  37m 23s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6141/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6141 |
| Optional Tests | dupname asflicense javac spotbugs checkstyle codespell detsecrets compile hadoopcheck hbaseanti spotless |
| uname | Linux 6cb5a04710d5 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / 20362ad2692cde45da2e12aa0348ac2bd29dd2fd |
| Default Java | Eclipse Adoptium-17.0.11+9 |
| spotless | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6141/1/artifact/yetus-general-check/output/patch-spotless.txt |
| Max. process+thread count | 85 (vs. ulimit of 30000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6141/1/console |
| versions | git=2.34.1 maven=3.9.8 spotbugs=4.7.3 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 26s |  |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  2s |  |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --author-ignore-list --blanks-eol-ignore-file --blanks-tabs-ignore-file --quick-hadoopcheck  |
|||| _ Prechecks _ |
|||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m 12s |  |  master passed  |
| +1 :green_heart: |  compile  |   0m 57s |  |  master passed  |
| +1 :green_heart: |  javadoc  |   0m 28s |  |  master passed  |
| +1 :green_heart: |  shadedjars  |   5m 47s |  |  branch has no errors when building our shaded downstream artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m  7s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 57s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 57s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 27s |  |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   5m 47s |  |  patch has no errors when building our shaded downstream artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 210m 40s |  |  hbase-server in the patch passed.  |
|  |   | 235m 43s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6141/1/artifact/yetus-jdk17-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6141 |
| Optional Tests | javac javadoc unit compile shadedjars |
| uname | Linux 7068fbe2dd64 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / 20362ad2692cde45da2e12aa0348ac2bd29dd2fd |
| Default Java | Eclipse Adoptium-17.0.11+9 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6141/1/testReport/ |
| Max. process+thread count | 5112 (vs. ulimit of 30000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6141/1/console |
| versions | git=2.34.1 maven=3.9.8 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/io/RangeStoreFileReader.java [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+
+  private boolean firstKeySeeked = false;
+
+  public RangeStoreFileReader(final ReaderContext context, final HFileInfo fileInfo,
+    final CacheConfig cacheConf, final byte[] startKey, final byte[] endKey, StoreFileInfo storeFileInfo,
+    final Configuration conf) throws IOException {
+    super(context, fileInfo, cacheConf, storeFileInfo, conf);
+    this.startKey = startKey;
+    this.endKey = endKey;
+    this.startCell = new KeyValue.KeyOnlyKeyValue(this.startKey, 0, this.startKey.length);
```
Comment (Apache9) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/io/RangeStoreFileReader.java:60]: At least you need to deal with startKey == null or endKey == null for the first range and last range?

Comment (chrajeshbabu) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/io/RangeStoreFileReader.java:60]: @Apache9 Handling in the next patch along with the test cases. Thanks

**Comment by chrajeshbabu** [comment]
Added test cases similar to halfstore file reader covering most of the happy path. Need to add few more scenarios of corner case test cases. Currently create RangeReference file but need to define the pattern for range reference and StoreFileInfo should automatically identify the Range Reference.
I have added few more TODOs like java docs etc..Will complete them.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 26s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  buf  |   0m  0s |  |  buf was not available.  |
| +0 :ok: |  buf  |   0m  0s |  |  buf was not available.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  |  Patch does not have any anti-patterns.  |
|||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 10s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m  9s |  |  master passed  |
| +1 :green_heart: |  compile  |   3m 38s |  |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 42s |  |  master passed  |
| +1 :green_heart: |  spotbugs  |   3m 51s |  |  master passed  |
| +1 :green_heart: |  spotless  |   0m 47s |  |  branch has no errors when running spotless:check.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 11s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m  4s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   3m 35s |  |  the patch passed  |
| +1 :green_heart: |  cc  |   3m 35s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   3m 35s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   0m 36s | [/results-checkstyle-hbase-server.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6141/2/artifact/yetus-general-check/output/results-checkstyle-hbase-server.txt) |  hbase-server: The patch generated 20 new + 0 unchanged - 0 fixed = 20 total (was 0)  |
| +1 :green_heart: |  spotbugs  |   4m  7s |  |  the patch passed  |
| +1 :green_heart: |  hadoopcheck  |  11m 35s |  |  Patch does not cause any errors with Hadoop 3.3.6 3.4.0.  |
| +1 :green_heart: |  hbaseprotoc  |   1m 22s |  |  the patch passed  |
| +1 :green_heart: |  spotless  |   0m 43s |  |  patch has no errors when running spotless:check.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 17s |  |  The patch does not generate ASF License warnings.  |
|  |   |  45m 43s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6141/2/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6141 |
| Optional Tests | dupname asflicense cc buflint bufcompat codespell detsecrets hbaseprotoc spotless javac spotbugs checkstyle compile hadoopcheck hbaseanti |
| uname | Linux 810b7668a3bb 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / 4cc18c7695a2b9ec105e51301b4273fd03fe0009 |
| Default Java | Eclipse Adoptium-17.0.11+9 |
| Max. process+thread count | 87 (vs. ulimit of 30000) |
| modules | C: hbase-protocol-shaded hbase-server U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6141/2/console |
| versions | git=2.34.1 maven=3.9.8 spotbugs=4.7.3 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 27s |  |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  3s |  |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --author-ignore-list --blanks-eol-ignore-file --blanks-tabs-ignore-file --quick-hadoopcheck  |
|||| _ Prechecks _ |
|||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 36s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m 26s |  |  master passed  |
| +1 :green_heart: |  compile  |   1m 30s |  |  master passed  |
| +1 :green_heart: |  javadoc  |   0m 35s |  |  master passed  |
| +1 :green_heart: |  shadedjars  |   5m 42s |  |  branch has no errors when building our shaded downstream artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 12s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m  8s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 30s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 30s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 36s |  |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   5m 41s |  |  patch has no errors when building our shaded downstream artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   0m 33s |  |  hbase-protocol-shaded in the patch passed.  |
| -1 :x: |  unit  | 210m 20s | [/patch-unit-hbase-server.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6141/2/artifact/yetus-jdk17-hadoop3-check/output/patch-unit-hbase-server.txt) |  hbase-server in the patch failed.  |
|  |   | 238m 27s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6141/2/artifact/yetus-jdk17-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6141 |
| Optional Tests | unit javac javadoc compile shadedjars |
| uname | Linux c4eac847de2a 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / 4cc18c7695a2b9ec105e51301b4273fd03fe0009 |
| Default Java | Eclipse Adoptium-17.0.11+9 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6141/2/testReport/ |
| Max. process+thread count | 5462 (vs. ulimit of 30000) |
| modules | C: hbase-protocol-shaded hbase-server U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6141/2/console |
| versions | git=2.34.1 maven=3.9.8 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.
